### PR TITLE
Limit concurrent workers in batch

### PR DESCRIPTION
**Pull request by calvinp0**
This pull request adds support for throttling the number of concurrently running array jobs ("max_concurrent") in pipe runs, allowing users to limit how many jobs are active at once on supported schedulers (Slurm, PBS, SGE, HTCondor, and OGE). It introduces a new `max_concurrent` parameter, updates submit script templates and metadata handling, and adds comprehensive tests to ensure correct behavior across schedulers and edge cases.

**Support for concurrent job throttling:**

* Added a `max_concurrent` parameter to `PipeRun`, allowing users to specify the maximum number of array jobs running simultaneously; this is validated and saved in run metadata, and is now loaded from disk for crash recovery. [[1]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8L54-R78) [[2]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8R87-R88) [[3]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8R123) [[4]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8R167) [[5]](diffhunk://#diff-2dddeabf54ef6b9b86c0f009bd7a15d448d902039d4955b0ed58e9cae372183dR309)
* The `submit_pipe_run` function and settings now pass `max_concurrent` through to `PipeRun`. [[1]](diffhunk://#diff-47f87a05e3c415332c1e336f66a45f96c3c1a663fc7898830314ec17c679a15bR157) [[2]](diffhunk://#diff-2dddeabf54ef6b9b86c0f009bd7a15d448d902039d4955b0ed58e9cae372183dR309)

**Scheduler template and logic updates:**

* The resource calculation and submit script rendering logic now compute and inject the correct throttling syntax for each supported scheduler (Slurm, PBS, SGE, OGE, HTCondor), including clamping and disabling as appropriate. [[1]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8L195-R248) [[2]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8L218-R267) [[3]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8L229-R279) [[4]](diffhunk://#diff-ea5fa52f1cc8d0ca6152fa77253652ee96b62199e5c588fb83918085a560e526L57-R60) [[5]](diffhunk://#diff-ea5fa52f1cc8d0ca6152fa77253652ee96b62199e5c588fb83918085a560e526L71-R72) [[6]](diffhunk://#diff-ea5fa52f1cc8d0ca6152fa77253652ee96b62199e5c588fb83918085a560e526L83-R86) [[7]](diffhunk://#diff-ea5fa52f1cc8d0ca6152fa77253652ee96b62199e5c588fb83918085a560e526L99-R100)

**Testing improvements:**

* Added extensive tests to `pipe_run_test.py` to verify throttling behavior, correct template rendering for each scheduler, value validation, persistence across reloads, and edge cases such as clamping and unthrottled runs. [[1]](diffhunk://#diff-9e3cf867c1cb5cbe551a8342ca37780b202f5810eb289f5c8ad670624987725bL137-R145) [[2]](diffhunk://#diff-9e3cf867c1cb5cbe551a8342ca37780b202f5810eb289f5c8ad670624987725bR176-R273)

**Internal code improvements:**

* Added a static method for validating `max_concurrent` values, ensuring only `None` or positive integers are accepted.
* Refactored internal APIs to use the new throttle-aware resource and template logic. [[1]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8L195-R248) [[2]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8L218-R267) [[3]](diffhunk://#diff-8bf6be4866906e63270e573fd1e9832168d26be71d7cb85abc6e60eea13d3bd8L229-R279)

These changes make job submission more robust and flexible for large-scale compute campaigns.

### CODE REVIEW

**Comment by alongd** [comment]
@calvinp0, can you take a look at the copilot comments, and explain the difference between `max_workers` and `max_concurrent`?

**Comment by calvinp0** [comment]
> @calvinp0, can you take a look at the copilot comments, and explain the difference between `max_workers` and `max_concurrent`?

Okay I will get back to you. My initial belief is that `max_workers` was how many jobs can be in one pipe before creating a second pipe job

**Comment by calvinp0** [comment]
> @calvinp0, can you take a look at the copilot comments, and explain the difference between `max_workers` and `max_concurrent`?

Okay its a lot more technical and complicated than I first understood.

So let's say we have 300 jobs to do.

And in the first case I do 8 max workers - max_concurrent is None.

8 workers spawn and share the 300 jobs via a claim pool (~37 each). They keep claiming until either the work runs out or they hit the queue's walltime. If walltime comes first, they die and  the leftovers sit PENDING - no one picks them up until we resubmit.

Now if we do 100 workers, and 8 concurrent. 

So 100 worker slots are queue, but the scheduler only lets 8 run at a time, the other 92 sit in the sched. queue with their wall time clock not yet started. If one of the active 8 dies (walltime, crash), the scheduler promotes a queued one, its fresh walltime begins and it picks up leftover PENDING tasks.

The difference I guess is max workers alone gives you a fixed lifetime budget. max concurrent with a alrger max workers gives you a rolling bench of reserves that rotate when an active worker dies.

Now overall I think the walltime case is not really applicable to us, depending on which queue you use, but I think for community usage it is good.

File: arc/job/pipe/pipe_run.py [hunk 1]
```diff
@@ -51,16 +52,31 @@ class PipeRun:
...
-        max_workers (int): Maximum number of concurrent array workers.
+        max_workers (int): Maximum total array worker slots (array size).
+        max_concurrent (Optional[int]): Max workers running simultaneously
+            (like PBS ``%N``). Must be ``None`` (unthrottled) or a positive
+            integer; ``0``, negatives, and non-integers raise ``ValueError``.
         max_attempts (int): Maximum retry attempts per task.
     """
 
+    @staticmethod
+    def _validate_max_concurrent(max_concurrent: Optional[int]) -> None:
```
Comment (alongd) [inline arc/job/pipe/pipe_run.py:63]: when rebasing, need to up the typehints to the Py14 style

File: arc/job/pipe/pipe_run.py [hunk 2]
```diff
@@ -192,12 +212,41 @@ def _submission_resources(self):
...
         array_size = min(self.max_workers, len(self.tasks)) if self.tasks else self.max_workers
-        return cpus, memory_mb, array_size
+        throttle = None
+        if self.max_concurrent is not None:
+            throttle = min(self.max_concurrent, array_size)
+        return cpus, memory_mb, array_size, throttle
+
+    def _render_throttle(self, array_size: int, throttle: Optional[int]) -> Dict[str, str]:
+        """
+        Render scheduler-specific array-range and extra-directives strings.
```
Comment (alongd) [inline arc/job/pipe/pipe_run.py:230]: please add the Google-style docstring, explaining `Args` and `Returns`

**Review by alongd — APPROVED** [review]
Looks good, just minor comments
